src/lib/y2storage/proposal/phys_vol_calculator.rb

```diff
@@ -0,0 +1,213 @@
+#!/usr/bin/env ruby
+#
+# encoding: utf-8
+
+# Copyright (c) [2015] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "y2storage/disk_size"
+require "byebug"
+
+module Y2Storage
+  class Proposal
+    # Class used by SpaceDistribution to find the best distribution of LVM
+    # physical volumes.
+    class PhysVolCalculator
+      # Initialize.
+      #
+      # @param all_spaces [Array<FreeDiskSpace>] Disk spaces that could
+      #     potentially contain physical volumes
+      # @param lvm_helper [Proposal::LvmHelper] contains information about the
+      #     LVM planned volumes and how to make space for them
+      def initialize(all_spaces, lvm_helper)
+        @all_spaces = all_spaces
+        @lvm_helper = lvm_helper
+      end
+
+      # Extended distribution that includes a planned volume for every
+      # physical volumes that would be necessary to fulfill the LVM requirements
+      #
+      # When using LVM, the number of possible distributions increases a lot.
+      # For every space in the disk we can decide to place an LVM PV or not.
+      # This method explores all the options and finds the best one.
+      #
+      # @param distribution [Proposal::SpaceDistribution] initial distribution
+      #     containing only the non-LVM volumes
+      #
+      # @return [Proposal::SpaceDistribution, nil] nil if it's impossible to
+      #     allocate all the needed physical volumes
+      def add_physical_volumes(distribution)
+        best = nil
+        all_spaces.permutation.each do |sorted_spaces|
```

Comment:
I'm not sure either :-) But I expect that to be solved in a upper layer. I mean. We search for all the spaces in `ProposalSettings#candidate_disks`. I hope we will have some pre-processing of the settings to make sure that when calling `DiskAnalyzer`, `PhysVolCalculator` and friends, `ProposalSettings#candidate_disks` does not contain hundreds of disks.

And we can even add some heuristic in the future to limit the options. For example, let's say that for GPT disks we don't need to explore all the possibilities because the distribution will be less tricky. All that can be added afterward.